File: emlun-scala/day07.scala
object Day07 extends App {
  val basePattern = raw"(\S+) \((\d+)\)(.*)".r

  case class RawProgram(name: String, weight: Int, children: Set[String])
  case class Program(name: String, weight: Int, children: Set[Program]) {
    lazy val childWeights: List[Int] = children.toList.map { _.totalWeight }
    lazy val totalWeight: Int = weight + childWeights.sum
    def isBalanced: Boolean = childWeights.toSet.size == 1
    def allChildrenBalanced: Boolean = children.forall { _.isBalanced }
  }

  def buildTree(root: RawProgram, allPrograms: Map[String, RawProgram]): Program =
    Program(
      name = root.name,
      weight = root.weight,
      children = root.children.map { child => buildTree(allPrograms(child), allPrograms) }
    )

  def parseChildren(rest: String): Set[String] =
    rest
      .stripPrefix(" -> ")
      .split(',')
      .map(_.trim)
      .filter(!_.isEmpty)
      .toSet

  def solveA(allPrograms: Set[RawProgram]): RawProgram = {
    val allDescendants: Set[String] = allPrograms.flatMap(_.children).toSet
    allPrograms.find({ p => !(allDescendants contains p.name) }).get
  }

  def solveB(root: Program): Option[Int] = {
    if (root.isBalanced)
      None
    else if (root.allChildrenBalanced) {
      val children = root.children
      val childWeights = root.childWeights

      val (goodWeights, badWeights) = childWeights.partition({ w => childWeights.count(_ == w) > 1 })
      val goodWeight = goodWeights.head
      val badWeight = badWeights.head

      val badProgram = children.find({ _.totalWeight == badWeight }).get

      Some(goodWeight - badProgram.childWeights.sum)
    }
    else
      root
        .children
        .flatMap(solveB)
        .headOption
  }

  val rawPrograms: Map[String, RawProgram] = (for {
    line <- io.Source.stdin.getLines
    if !line.trim.isEmpty
    program = {
      line.trim match {
      case basePattern(name, weight, rest) => RawProgram(name, weight.toInt, parseChildren(rest))
    }}
  } yield (program.name, program)).toMap

  val root = solveA(rawPrograms.values.toSet)

  println(s"A: ${root.name}")
  println(s"B: ${solveB(buildTree(root, rawPrograms)).get}")
}

File: plilja-python3/day07/day07.py
import re
import sys
from collections import namedtuple, Counter


Program = namedtuple('Program', 'name children weight')


def read_input():
    programs = []
    for line in sys.stdin.readlines():
        program, weight, rest = re.match(r'(.+) \((\d+)\)(.*)', line).groups()
        if '->' in rest:
            children = rest[4:].strip().split(', ')
        else:
            children = []
        programs += [Program(program, children, int(weight))]
    return programs


def step1(programs):
    children = set()
    for program in programs:
        children |= set(program.children)
    for program in programs:
        if program.name not in children:
            return program
    return None


def step2(programs, root):
    weight_cache = {}
    namelookup = dict([(p.name, p) for p in programs])


    def weight(program):
        if program.name in weight_cache:
            return weight_cache[program.name]
        else:
            w = sum([weight(namelookup[c]) for c in program.children])
            w += program.weight
            weight_cache[program.name] = w
            return w


    def is_balanced(program):
        for c in program.children:
            if not is_balanced(namelookup[c]):
                return False
        s = set([weight(namelookup[c]) for c in program.children])
        return len(s) <= 1


    def go(program):
        for c in program.children:
            if not is_balanced(namelookup[c]):
                return go(namelookup[c])
        counter = Counter([weight(namelookup[c]) for c in program.children])
        target = counter.most_common()[0][0]
        actual = counter.most_common()[1][0]
        for c in program.children:
            if weight(namelookup[c]) == actual:
                return namelookup[c].weight + target - actual
        return None

        
    return go(root)


programs = read_input()
root = step1(programs)
print(root.name)
print(step2(programs, root))
